reconcile/openshift_resources.py

```diff
@@ -0,0 +1,271 @@
+    def __init__(self, body):
+        super(OpenshiftResource, self).__init__(
+            body, QONTRACT_INTEGRATION, QONTRACT_INTEGRATION_VERSION
+        )
+
+
+class ResourceInventory(object):
+    def __init__(self):
+        self._dict = {}
+
+    def initialize_resource_type(self, cluster, namespace, resource_type):
+        self._dict.setdefault(cluster, {})
+        self._dict[cluster].setdefault(namespace, {})
+        self._dict[cluster][namespace].setdefault(resource_type, {
+            'current': {},
+            'desired': {}
+        })
+
+    def add_desired(self, cluster, namespace, resource_type, name, value):
+        desired = self._dict[cluster][namespace][resource_type]['desired']
+        desired[name] = value
+
+    def add_current(self, cluster, namespace, resource_type, name, value):
+        current = self._dict[cluster][namespace][resource_type]['current']
+        current[name] = value
+
+    def iterate(self):
+        for cluster in self._dict.keys():
+            for namespace in self._dict[cluster].keys():
+                for resource_type in self._dict[cluster][namespace].keys():
+                    data = self._dict[cluster][namespace][resource_type]
+                    yield (cluster, namespace, resource_type, data)
+
+
+def fetch_provider_resource(path):
+    gqlapi = gql.get_api()
+
+    # get resource data
+    try:
+        resource = gqlapi.get_resource(path)
+    except gql.GqlApiError as e:
+        raise FetchResourceError(e.message)
+
+    try:
+        resource['body'] = anymarkup.parse(
+            resource['content'],
+            force_types=None
+        )
+    except anymarkup.AnyMarkupError:
+        e_msg = "Could not parse data. Skipping resource: {}"
+        raise FetchResourceError(e_msg.format(path))
+
+    openshift_resource = OpenshiftResource(resource['body'])
+
+    # verify valid k8s object
+    try:
+        openshift_resource.run_assert()
+    except TypeError:
+        e_msg = "Invalid data (TypeError). Skipping resource: {}"
+        raise FetchResourceError(e_msg.format(path))
+    except KeyError:
+        e_msg = "Invalid data (KeyError). Skipping resource: {}"
+        raise FetchResourceError(e_msg.format(path))
+
+    return openshift_resource
+
+
+def fetch_data(namespaces_query):
+    ri = ResourceInventory()
+    oc_map = {}
+    errors = []
+
+    for namespace_info in namespaces_query:
+        namespace = namespace_info['name']
+
+        # Skip if namespace has no managedResourceTypes
+        managed_types = namespace_info.get('managedResourceTypes')
+        if not managed_types:
+            continue
+
+        # Obtain `oc` client
+        cluster_info = namespace_info['cluster']
+        cluster = cluster_info['name']
+        if oc_map.get(cluster) is None:
+            at = cluster_info.get('automationToken')
+
+            # Skip if cluster has no automationToken
+            if at is None:
+                continue
```

Comment:
maybe it worth to rise WARN or something here because it might be tricky to debug for new cluster why it does nothing. WDYT?

Reply:
ah, I see what you mean. If it has managedRoles or managedResourceTypes but no automationToken, it should definitely fail, yeah. I think we should also fail the PR.

Making the change.